### DISCUSSION TITLE
Support a different Prior per channel

### DISCUSSION THREAD

**Discussion by luiztauffer** [pymc-labs/pymc-marketing#1267] [Ideas]
I would like to be able to define different prior functions for each channel type (and eventually also control parameters). I think the API could be changed in two different ways:

1. addressing channels by name (preferable):
```python
model_config = {  
    "saturation_beta": {  
        “channel_1”: {"dist": "HalfNormal",  "kwargs": {"sigma": 0.2} },  
        “channel_2”: {"dist": "Normal",  "kwargs": {"mu": 1, "sigma": 0.3} },  
    },
...
}  
```

2. addressing channel as a list (not ideal, but perhaps easier to implement?):
```python
model_config = {  
    "saturation_beta": [
       {"dist": "HalfNormal",  "kwargs": {"sigma": 0.2} },  
       {"dist": "Normal",  "kwargs": {"mu": 1, "sigma": 0.3} },  
    ],
...
}  
```

This would be relevant in at least two scenarios:
- when working on a problem where different channels are expected to have quite distinct properties, e.g. temporal adstock profiles.
- if you want to have hierarchical grouping by each channel types (e.g. all channel_1 from different experiments) instead of grouping all channel types in the same hierarchy

Are there plans to implement something like this?
@Paulo-MT2 and I have been using pymc-marketing a lot recently and playing around with changes to it, so we could definitely help building that into the package, if the maintainers could help giving us guidance on where to make the modifications.

**Comment by williambdean** [comment]
For Normal and HalfNormal, try TruncatedNormal with vector of paramters. -np.inf will be same as normal

Other mixtures of different distributions are not supported at the moment via Prior class

**Comment by luiztauffer** [comment]
@wd60622 do you mean this should work?

```python
model_config = {  
    "saturation_beta": {
        "dist": "Normal",
        "kwargs": {
            "mu": [0.2, 0.6],
            "sigma": [0.1, 0.3],
        }
    },  
} 
```
where `mu` and `sigma` are vectors with the same size as the number of channels?

Can you please clarify if this is creating:

1. a different prior per channel
2. a prior that is a mixture of distributions, but is still the same prior for all channels

**Reply by williambdean** [reply]
So the dictionary syntax will still work but that is the old format (from a few versions ago). The `Prior` class is now used.

That would work if you had two channels which you'd like to be Normal. 

I meant something like this:

```python
import numpy as np

from pymc_marketing.prior import Prior

# HalfNormal, HalfNormal, Normal via TruncatedNormal
saturation_beta = Prior(
    "TruncatedNormal",
    mu=[0, 1, 2],
    sigma=[1, 1.5, 2.5],
    lower=[0, 0, -np.inf],
    dims="channel",
)

model_config = {
    "saturation_beta": saturation_beta,
    ..., 
}
```

The priors can be visualized: 

```python
import matplotlib.pyplot as plt

coords = {"channel": ["A", "B", "C"]}
samples = saturation_beta.sample_prior(coords=coords)

samples["var"].to_series().unstack().hist(sharex=True)
plt.show()
```
![half-normal](https://github.com/user-attachments/assets/07089a80-e6a5-400c-adbe-fc96cde95735)

**Reply by luiztauffer** [reply]
thank you for the example, @wd60622 ! This makes sense now, we'll try to use it like that.
